From 925153dae70cfd38c58c774c49c497bf675ce93d Mon Sep 17 00:00:00 2001
From: Andy Yan <andy.yan@rock-chips.com>
Date: Thu, 12 Nov 2020 09:24:22 +0800
Subject: [PATCH] drm/rockchip: vop2: Convert 10 bits output mode to 8 bits for
 vp max bpc is 8

Not all the video ports support 10 bit output, so
we need do a convert here.

Change-Id: I6dbb0c23e9cc1ceaf295e2771606705dacd37994
Signed-off-by: Andy Yan <andy.yan@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 8 +++++++-
 drivers/gpu/drm/rockchip/rockchip_vop2_reg.c | 1 +
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index 42efb01ff671..b6de49fc86e1 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -2504,6 +2504,7 @@ static void vop2_crtc_atomic_enable(struct drm_crtc *crtc, struct drm_crtc_state
 	u16 vact_end = vact_st + vdisplay;
 	bool interlaced = !!(adjusted_mode->flags & DRM_MODE_FLAG_INTERLACE);
 	int sys_status = SYS_STATUS_LCDC0;
+	uint8_t out_mode;
 	int for_ddr_freq = 0;
 	bool dclk_inv;
 	int act_end;
@@ -2630,7 +2631,12 @@ static void vop2_crtc_atomic_enable(struct drm_crtc *crtc, struct drm_crtc_state
 		VOP_CTRL_SET(vop2, hdmi_dclk_pol, 1);
 	}
 
-	VOP_MODULE_SET(vop2, vp, out_mode, vcstate->output_mode);
+	if (vcstate->output_mode == ROCKCHIP_OUT_MODE_AAAA &&
+	    !(vp_data->feature & VOP_FEATURE_OUTPUT_10BIT))
+		out_mode = ROCKCHIP_OUT_MODE_P888;
+	else
+		out_mode = vcstate->output_mode;
+	VOP_MODULE_SET(vop2, vp, out_mode, out_mode);
 	VOP_MODULE_SET(vop2, vp, overlay_mode, is_yuv_output(vcstate->bus_format));
 
 	vop2_dither_setup(crtc);
diff --git a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
index f7dcbc8b2a73..9775ceb9896d 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
@@ -466,6 +466,7 @@ static const struct vop2_video_port_data rk3568_vop_video_ports[] = {
 	{
 	 .id = 0,
 	 .soc_id = 0x3568,
+	 .feature = VOP_FEATURE_OUTPUT_10BIT,
 	 .pre_scan_max_dly = { 40, 49, 33, 42 },
 	 .intr = &rk3568_vp0_intr,
 	 .hdr_table = &rk3568_vop_hdr_table,
-- 
2.35.3

